File: 2020/06_CustomCustoms/test_group.py
# ======================================================================
# Custom Customs
#   Advent of Code 2020 Day 06 -- Eric Wastl -- https://adventofcode.com
#
# Python implementation by Dr. Dean Earl Wright III
# ======================================================================

# ======================================================================
#                        t e s t _ g r o u p . p y
# ======================================================================
"Test set of answers for Advent of Code 2020 day 06, Custom Customs"

# ----------------------------------------------------------------------
#                                                                 import
# ----------------------------------------------------------------------
import unittest

import group

# ----------------------------------------------------------------------
#                                                              constants
# ----------------------------------------------------------------------
EXAMPLE_TEXT = "abcx abcy abcz"
EXAMPLES = [
    {'text': 'abc', 'num': 3, 'num2': 3},
    {'text': 'a b c', 'num': 3, 'num2': 0},
    {'text': 'ab ac', 'num': 3, 'num2': 1},
    {'text': 'a a a a', 'num': 1, 'num2': 1},
    {'text': 'b', 'num': 1, 'num2': 1},
]

# ======================================================================
#                                                              TestGroup
# ======================================================================


class TestGroup(unittest.TestCase):  # pylint: disable=R0904
    "Test Group object"

    def test_empty_init(self):
        "Test the default Group creation"

        # 1. Create default Group object
        myobj = group.Group()

        # 2. Make sure it has the default values
        self.assertEqual(myobj.part2, False)
        self.assertEqual(myobj.text, None)
        self.assertEqual(len(myobj.answers), 0)

    def test_text_init(self):
        "Test the Group object creation from text"

        # 1. Create Group object from text
        myobj = group.Group(text=EXAMPLE_TEXT)

        # 2. Make sure it has the expected values
        self.assertEqual(myobj.part2, False)
        self.assertEqual(len(myobj.text), 14)
        self.assertEqual(len(myobj.answers), 6)

    def test_example(self):
        "Test part one example of Group object"

        # 1. Loop for all the examples
        for example in EXAMPLES:

            # 2. Create Group object from text
            myobj = group.Group(text=example['text'])

            # 2. Check the number of yes aanswers
            self.assertEqual(len(myobj.answers), example['num'])

    def test_example2(self):
        "Test part two example of Group object"

        # 1. Loop for all the examples
        for example in EXAMPLES:

            # 2. Create Group object from text
            myobj = group.Group(text=example['text'], part2=True)

            # 2. Check the number of yes aanswers
            self.assertEqual(len(myobj.answers), example['num2'])


# ----------------------------------------------------------------------
#                                                  module initialization
# ----------------------------------------------------------------------
if __name__ == '__main__':
    pass

# ======================================================================
# end                     t e s t _ g r o u p . p y                  end
# ======================================================================

File: 2020/06_CustomCustoms/test_group.lua
-- ======================================================================
-- Custom Customs
--   Advent of Code 2020 Day 06 -- Eric Wastl -- https://adventofcode.com
--
-- lua implementation by Dr. Dean Earl Wright III
-- ======================================================================

-- ======================================================================
--                    t e s t _ g r o u p . l u a
-- ======================================================================
-- Test Group for Advent of Code 2020 day 06, Custom Customs

-- ----------------------------------------------------------------------
--                                                                require
-- ----------------------------------------------------------------------
local luaunit = require('luaunit')

local group = require('group')

-- ----------------------------------------------------------------------
--                                                              constants
-- ----------------------------------------------------------------------
local EXAMPLE_TEXT = { "abcx", "abcy", "abcz" }

-- ======================================================================
--                                                              TestGroup
-- ======================================================================

function test_empty_init()
  -- Test the default Group creation

  -- 1. Create default Group object
  local myobj = group:Group()

  -- 2. Make sure it has the default values
  luaunit.assertEquals(myobj.part2, false)
  luaunit.assertEquals(#myobj.text, 0)
  luaunit.assertEquals(#myobj.questions, 0)

end

function test_text_init()
  -- Test the Group object creation from text

  -- 1. Create Groups object from text
  local myobj = group:Group({text=EXAMPLE_TEXT})

  -- 2. Make sure it has the expected values
  luaunit.assertEquals(myobj.part2, false)
  luaunit.assertEquals(#myobj.text, 3)
  
  -- 3. Check methods
  luaunit.assertEquals(myobj:count_yes(), 6)
  

end

function test_part2()
  -- Test the Group object creation from text for part2

  -- 1. Create Groups object from text
  local myobj = group:Group({part2=true, text=EXAMPLE_TEXT})

  -- 2. Make sure it has the expected values
  luaunit.assertEquals(myobj.part2, true)
  luaunit.assertEquals(#myobj.text, 3)
  
  -- 3. Check methods
  luaunit.assertEquals(myobj:count_yes(), 3)
  

end

-- ----------------------------------------------------------------------
--                                                  module initialization
-- ----------------------------------------------------------------------
os.exit( luaunit.LuaUnit.run() )

-- ======================================================================
-- end                    t e s t _ g r o u p . p y                   end
-- ======================================================================
